### PR TITLE
Move rain sources to new tend spec

### PR DESCRIPTION
**Pull request by charleskawczynski**
This PR:
 - Adds a `source` validation assert: `@assert !any(isa.(source, Tuple))` to the `AtmosModel` to break early if we've incorrectly specified sources (e.g., not splatting them). This addresses #1779 but only for the AtmosModel.
 - Renames `Rain` to `RainModel`, and adds `Rain <: PrognosticVariable` and `Snow <: PrognosticVariable`.
 - Moves the `RemovePrecipitation` to the new tendency specification
 - Moves the `Rain_1M` to the new tendency specification

Here's the new table with `Dycoms` using non-equilibrium model with `RainModel`:
```
PDE: ∂_t Y_i + (∇•F_1(Y))_i + (∇•F_2(Y,G)))_i = (S(Y,G))_i
┌────────────────┬────────────────────────────┬─────────────────────────────────────────┬───────────┐
│       Equation │           Flux{FirstOrder} │                       Flux{SecondOrder} │    Source │
│          (Y_i) │                      (F_1) │                                   (F_2) │       (S) │
├────────────────┼────────────────────────────┼─────────────────────────────────────────┼───────────┤
│           Mass │                   (Advect) │                     (MoistureDiffusion) │ (Rain_1M) │
│       Momentum │ (Advect, PressureGradient) │                         (ViscousStress) │        () │
│         Energy │         (Advect, Pressure) │ (ViscousProduction, EnthalpyProduction) │ (Rain_1M) │
│  TotalMoisture │                   (Advect) │                                      () │ (Rain_1M) │
│ LiquidMoisture │                   (Advect) │                                      () │ (Rain_1M) │
│    IceMoisture │                   (Advect) │                                      () │        () │
│           Rain │                         () │                                      () │ (Rain_1M) │
└────────────────┴────────────────────────────┴─────────────────────────────────────────┴───────────┘
```

This PR builds on/supersedes #1786.

### Description

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by charleskawczynski** [comment]
bors try

**Comment by charleskawczynski** [comment]
bors try

**Comment by charleskawczynski** [comment]
Hmm, a bunch of things are breaking, we might need to break this PR up

**Comment by charleskawczynski** [comment]
bors try

File: src/Atmos/Model/multiphysics_types.jl [hunk 1]
```diff
@@ -19,3 +22,100 @@ struct PressureGradient{PV <: Momentum} <: TendencyDef{Flux{FirstOrder}, PV} end
...
+       TendencyDef{Source, PV}
+    " Set to true if using qc based threshold"
+    use_qc_thr::Bool
+end
+
+RemovePrecipitation(b::Bool) = (
+    RemovePrecipitation{Mass}(b),
+    RemovePrecipitation{Energy}(b),
+    RemovePrecipitation{TotalMoisture}(b),
+)
```
Comment (trontrytel) [inline src/Atmos/Model/multiphysics_types.jl:46]: I think this is nice. It clearly shows what state variables are affected by each source term.

**Review by trontrytel — APPROVED** [review]
LGTM! Lets see the results in nightly qa-microphysics!

**Pull request merged**